### ISSUE TITLE
Make it possible to dump benchmark diff report in JSON format

### ISSUE THREAD

**Issue by JBakamovic** [google/benchmark#737]
TL;DR Be able to dump diff report of two benchmarks into the JSON file and not only on stdout as it is right now. E.g. `compare.py -d bench_diff.json benchmarks bench1.json bench2.json`

Rationale behind this functionality is to be able to manipulate the performance data directly which is much more powerful than simply printing the data onto the stdout. Some other higher-level scripts could then consume this API to make use of the data in whatever way it wants to.

One example could be detection of performance regressions integrated tightly into the CI/CD pipeline. Another one would be a plotter which would be displaying how performance data has been changing through different commits/releases/etc. You get the idea ...

Current implementation of `compare.py` does not have this functionality so I've prepared a small patch which I will send shortly afterwards.

Let me know what you think about this feature.

**Comment by LebedevRI** [comment]
> TL;DR Be able to dump diff report of two benchmarks into the JSON file and not only on stdout as it is right now. E.g. `compare.py -d bench_diff.json benchmarks bench1.json bench2.json`
> 
> Rationale behind this functionality is to be able to manipulate the performance data directly which is much more powerful than simply printing the data onto the stdout. Some other higher-level scripts could then consume this API to make use of the data in whatever way it wants to.

I don't dislike this in general, but right now it is a just simple script.
i'm really uneasy about making **any** API/json stability guarantees.

> One example could be detection of performance regressions integrated tightly into the CI/CD pipeline. Another one would be a plotter which would be displaying how performance data has been changing through different commits/releases/etc. You get the idea ...

This really should be done by something much more specialized, like llvm lnt..

> Current implementation of `compare.py` does not have this functionality so I've prepared a small patch which I will send shortly afterwards.
> 
> Let me know what you think about this feature.

**Comment by JBakamovic** [comment]
> I don't dislike this in general, but right now it is a just simple script.

Yes it is super simple right now but why wouldn't it provide functionality that basically anyone can have benefit of? And the change is a really simple one ...

> i'm really uneasy about making any API/json stability guarantees.

But we already have a JSON format output defined when dumping the benchmark results. Outputting the diff report to JSON would be clearly built upon it.

> This really should be done by something much more specialized, like llvm lnt..

Sorry but I fail to see what does LLVM has to do with this topic?

**Comment by LebedevRI** [comment]
> > I don't dislike this in general, but right now it is a just simple script.
> 
> Yes it is super simple right now but why wouldn't it provide functionality that basically anyone can have benefit of? And the change is a really simple one ...
> 
> > i'm really uneasy about making any API/json stability guarantees.
> 
> But we already have a JSON format output defined when dumping the benchmark results. Outputting the diff report to JSON would be clearly built upon it.

Eeh, these two phrases are just one phrase, i meant to use comma, not full stop.
It should have been:
> I don't dislike this in general, but right now it is a just simple script,
> i'm really uneasy about making any API/json stability guarantees.

> But we already have a JSON format output defined when dumping the benchmark results. Outputting the diff report to JSON would be clearly built upon it.

So can that output be fed back into the `compare.py` script?

> > This really should be done by something much more specialized, like llvm lnt..
> 
> Sorry but I fail to see what does LLVM has to do with this topic? Proposed functionality is a generic one that any project could make use of it without any other dependencies.

The **usual** CI is unfit to do any meaningful performance measurements.
It will be **extremely** noisy, because the hardware will be shared.
Thus i'm completely questioning the reasoning here.

**Comment by JBakamovic** [comment]
> So can that output be fed back into the compare.py script?

No, I have not envisioned that use-case. Not sure what `compare.py` could do in that particular case but fail.

> The usual CI is unfit to do any meaningful performance measurements.

CI is just one example that I gave. Different projects will have different needs or means of post-processing the performance data. The most simple one is a plotter.

> It will be extremely noisy, because the hardware will be shared.

Yes, this is a CI challenge but can be tackled by having a dedicated no-shared node which will do only performance tests one at a time. Having reproducible and continuous performance regressions detection is a very important aspect in developing high-quality software, IMO. Or in other words, being able to detect regressions automatically and early during the development cycle.

**Comment by JBakamovic** [comment]
> Sure, just want to point out that then it will no longer be advertisable as "reliable to use by any project with no special setup.".

Agreed, but understanding performance implications has never been a trivial task. Even with this framework you still have to know what you're doing regardless of dump-to-json functionality. If you accept the PR
